File: delegateadapter/src/main/java/com/livermor/delegateadapter/delegate/diff/DiffUtilCallback.java
package com.livermor.delegateadapter.delegate.diff;


import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

/**
 * @author dumchev on 16.11.17.
 */
public class DiffUtilCallback extends DiffUtil.Callback {

    private final List<? extends DiffUtilItem> oldList;
    private final List<? extends DiffUtilItem> newList;

    public DiffUtilCallback(@NonNull List<? extends DiffUtilItem> oldList,
                     @NonNull List<? extends DiffUtilItem> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).id().equals(newList.get(newItemPosition).id());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).content().equals(newList.get(newItemPosition).content());
    }
}

File: delegateadapter/src/main/java/com/livermor/delegateadapter/delegate/DiffUtilCallback.kt
package com.livermor.delegateadapter.delegate

import androidx.recyclerview.widget.DiffUtil

/**
 * @author dumchev on 16.11.17.
 */
class DiffUtilCallback(
    private val oldState: AdaptersState,
    private val newState: AdaptersState
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = oldState.data.size
    override fun getNewListSize(): Int = newState.data.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
        oldState.itemId(oldItemPosition) == newState.itemId(newItemPosition)

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
        oldState.itemContent(oldItemPosition) == newState.itemContent(newItemPosition)

    private fun AdaptersState.itemId(position: Int): Any =
        getAdapterByItemPosition(position).itemId(data[position])

    private fun AdaptersState.itemContent(position: Int): Any =
        getAdapterByItemPosition(position).itemContent(data[position])
}